packages/flutter/lib/src/rendering/sliver_list.dart

```diff
@@ -38,24 +47,171 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
+        // Remove the cached extent if widget's key changed.
+        _cachedItemExtents.remove(index);
+        return false;
+      }
+    }
+    return false;
+  }
+
+  RenderBox? _obtainOneLeadingChild() {
+    final int leadingIndex = indexOf(firstChild!) - 1;
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(leadingIndex);
+    final RenderBox? leadingChild = insertAndLayoutLeadingChild(
+      _childConstraints,
+      parentUsesSize: true,
+      layoutImmediate: layoutImmediate,
+    );
+    if (enableItemExtentsCaching && leadingChild != null) {
+      assert(leadingChild == firstChild);
+      if (layoutImmediate) {
+        _cacheItemExtent(firstChild!);
+      } else {
+        _childrenWithoutLayout.add(leadingIndex);
+      }
+    }
+    return leadingChild;
+  }
+
+  RenderBox? _obtainOneTrailingChild(RenderBox after) {
+    final int trailingIndex = indexOf(after) + 1;
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(trailingIndex);
+    final RenderBox? trailingChild = insertAndLayoutChild(
+      _childConstraints,
+      after: after,
+      parentUsesSize: true,
+      layoutImmediate: layoutImmediate,
+    );
+    if (enableItemExtentsCaching && trailingChild != null) {
+      if (layoutImmediate) {
+        _cacheItemExtent(trailingChild);
+      } else {
+        _childrenWithoutLayout.add(trailingIndex);
+      }
+    }
+    return trailingChild;
+  }
+
+  void _processChildLayout(RenderBox child) {
+    final int childIndex = indexOf(child);
+    final bool layoutImmediate = !enableItemExtentsCaching || !_canUseCachedExtent(childIndex);
+    if (layoutImmediate) {
+      child.layout(_childConstraints, parentUsesSize: true);
+      if (enableItemExtentsCaching) {
+        _cacheItemExtent(child);
+      }
+    } else {
+      _childrenWithoutLayout.add(childIndex);
+    }
+  }
+
+  void _performLoadingScopeChildrenLayoutIfNeeded() {
+    if (!enableItemExtentsCaching) {
+      return;
+    }
+    for (RenderBox? child = firstChild; child != null ; child = childAfter(child)) {
+      final int index = indexOf(child);
+      if (_childrenWithoutLayout.contains(index)) {
+        child.layout(_childConstraints, parentUsesSize: true);
+        // Throw if the real extent differs from the cached value.
+        assert(() {
+          if (super.paintExtentOf(child!) != _cachedItemExtents[index]!.mainAxisExtent) {
+            throw FlutterError(
+              'The item(index: $index) main axis extent has changed after caching, if this is '
+              'expected, change its Key to update the cached value.',
+            );
+          }
+          return true;
+        }());
+      }
+    }
+  }
+
+  void _setup() {
+    _childrenWithoutLayout.clear();
+    _childConstraints = constraints.asBoxConstraints();
+    // Clear the cached extents if [crossAxisExtent] or [axis] changed.
+    if (constraints.crossAxisExtent != _lastCrossAxisExtent ||
```

Comment:
viewportMainAxisExtent would probably clear the cache too.

Reply:
Generally, the layout of items in the main axis direction is not constrained, so the cache is not cleared. At the same time, developers can explicitly clear the cache data through `invalidCachedItemExtents`.
Maybe both behaviors are okay and I can accept both.

Reply:
> Generally, the layout of items in the main axis direction is not constrained

Totally! You are right. :)
Some folks size/position their children based on the viewport dimensions, especially not that we provide that information in the itemExtentBuilder feature you added recently. So it would probably be good to use a change in the viewport dimension as a cache invalidator.